design/one-pager-merge-jets-proposal.md

```diff
@@ -0,0 +1,370 @@
+would see a clear path forward and we'd not waste any efforts by the community.
+
+Note that we are making two decisions here:
+* There will be a single provider.
+* The single provider will be the classic one and we'll add Terrajet CRDs on
+  top.
+
+There are several reasons to choose this approach instead of having providers
+separated by whether they call APIs directly or use Terrajet. Let's take a look
+at them one by one.
+
+### Dependence on Terraform Provider
+
+The generic Terrajet controller calls Terraform CLI which in turn calls its
+Terraform provider to operate. Each CRD has its own completely independent
+reconciler thread configured to work with the schema of that resource.
+
+This dependence causes the Crossplane provider to be exposed to the bugs of the
+underlying Terraform provider. While we strive to fix the bug there first, it's
+not always feasible for us to land that change in a timely manner. In addition,
+there are cases where we are locked to a version with a bug where downgrading or
+upgrading is not an option because of another bug or behavior change.
+
+With the mixed provider, we can change the underlying implementation to a
+hand-written controller whenever we face such a bug. With the Kubernetes API
+conversion tooling, we can manage the possible API changes as well. This would
+also mean that over time, we'd have less and less Terraform dependence in our
+ecosystem.
+
+### Confusion for Users and Maintainers
+
+Terraform is a great tool but in an ideal state, we'd like the cloud vendors to
+step in and maintain those providers. They may choose to keep using their TF
+provider but it's unlikely; AWS does not maintain their TF provider, Azure uses
+ARM in their TF provider and they've shown interest in generating Crossplane
+based on ARM and Google is using Magic Modules and DCL to generate TF provider,
+so it's likely that they will choose to do that for their Crossplane provider as
+well. [Provider Strategy doc][strategy-doc] goes into more detail about these
+tools. In summary, Terrajet will be a temporary solution that we put in place to
+get more coverage and at some point we'll need to discuss how to change the
+implementation to those tools.
+
+With that in mind, it's a hard choice to invest in separate Jet-based providers
+knowing that they will, even though after years, be deprecated at one point and
+users will have to migrate. At the same time, classic providers are very far
+from having the same coverage level. This creates confusion and reluctance to
+use either of the providers, which also results in decreased pace of
+contributions because of the divergence of efforts.
+
+With a single mixed provider, the story is much clearer; we fill the gaps with
+Terrajet and over time the implementation of the resources will change to be
+more _native_; be it cloud vendors using their own approaches or handwritten API
+calls community writes that may be necessary to overcome a bug in TF provider.
```

Comment:
> I'm betting option 2 will waste hundreds of contributor hours building MRs a vendor will eventually throw away.

Could you elaborate on this? I don't really set a goal to converge on classic; either vendors will convert them to whatever they'd like or we will convert but only when we **have to** due to a bug - no explicit goal of getting to full classic. So, we'll add new CRDs to the classic providers via Terrajet and that's pretty much it. We'll do any sort of conversion only if it's deemed necessary by its maintainers; which could be the vendor or community maintainers. No one will build new native MRs, Terrajet will be the first choice.

If I understood correctly, your preference of option 3 suggests that we set a goal to convert SDK implementations to Jet. Assuming the conversion webhooks are in place, hence we're able to make breaking API changes, this would mean replacing the controllers of 127 CRDs in provider-aws. The costs of doing this in my mind are:
* Risk. Each replacement of controller, be it in either way, carries some amount of risk of bugs. Even though you get the API conversion right, controller behavior may not exactly match. We'd have to be careful about this risk for especially v1beta1 resources.
  * The proposal suggests that we do this only if we have to due to a bug or vendor decides to do so.
* Effort. We can use Terrajet to generate these CRDs right away but the schema conversion functions will have to be handwritten and we need to test each of them to make sure existing users are not affected. Especially v1beta1 ones will make the testing more cumbersome.
  * The proposal suggests to do that only if we can't fix a bug in a Jet resource and its direction would be Jet -> classic, so we're going from less transparency to full control which means we have more tooling available to make the controller behave exactly like its Jet variant.
* Community backlash. These resource have been in place since the existence of Crossplane and many people in the community contributed to them in one way or another. Replacing the controller code with a temporary solution which depends on Terraform may not be well-received.

> As we've identified, each of the big three vendors seems to have their own preferred way to build providers (e.g. AWS Cloud Control API, Azure ARM, GCP DCL).

I agree that when the vendors step in to maintain their providers, they'll try to make the best decision to make it more maintainable for themselves. However, this doesn't directly mean that it will result in a different provider. For example, GCP DCL does not have full coverage, so they use it only for a portion of the provider. As a counter example, TF folks decided to start another provider for CC.

IMO, the reason they will step in will be the popularity of the community-maintained providers and they would possibly decide to be maintainers on those repos and change the implementation to their tools by using conversion mechanisms we have in place.

In my opinion, the best we can do is to strive to make such replacements as less risky as possible so that they take part in the existing provider. If they decide to bootstrap another one, like CC, then that's also perfectly fine because in the meantime, the community will have coverage **and** maturity option already and they can decide to move to that provider whenever they'd like to, which is similar to how [AWS CC explains](https://github.com/hashicorp/terraform-provider-awscc/blob/main/contributing/FAQ.md#which-provider-should-i-use) their two provider approach.

Reply:
@negz I just pushed an update adding the tangible steps we'll take to merge the providers and also the guidelines to follow when you add or convert a controller from one approach to another. Let me know if you'd like to see more detailed guidelines or want to change them.

> I agree with these costs, but I think they are worth paying to provide users and contributors a more consistent experience going forward.

Minding the costs, I still don't really think we should set a goal of converting everything to Terrajet but tried to list a few cases where you'd like to do that and it's totally fine, i.e. the golden standard is not native, it's whatever best for that specific resource. Though I believe we will end up with more and more converted-to-Jet controllers in the long run, especially with the APIs that are really cumbersome to manage like DynamoDB or CloudFront Distribution; but not much for EKS cluster or VPC.

Reply:
I got a little confused with this discussion. Before this proposal, I had in my mind that hand-written MR implementation was preferred over TJ-generated ones. And TJ was only serving as a fast way to increase coverage for providers, but in the future, we should strive to have all MRs to be hand-written.

Now I'm getting a little confused, because if the true goal is to have vendors to maintain their own providers (and they can simply throw away the community-one to create another implementation for themselves), contributors would end up "wasting" a lot of time implementing hand-written MRs/transforming TJ to classic. And this wouldn't happen with TJ-generated implementation, because it is significantly faster/effortless then the hand-written option.

So I think it's important to set a goal here, just to be clear for contributors what should we do until vendors owns the maintainership of their providers. 

When we need to add a new resource, its clear to me what we have to do when deciding about TJ/classic.

But how do we should **preferably** iterate on these resources? Trying to be more clear in my question, if we have all providers with 100% coverage and in a stable version in the future, what should this provider looks like, 99% TJ or 99% classic? Which is the best/preferable implementation for MRs?

Reply:
> But how do we should preferably iterate on these resources? Trying to be more clear in my question, if we have all providers with 100% coverage and in a stable version in the future, what should this provider looks like, 99% TJ or 99% classic? Which is the best/preferable implementation for MRs?

I think reducing the whole thing to which one is better does not serve well for all resources. Even though we standardize the cloud provider APIs behind XRM, each (service * cloud) API combination has its own quirks and some of them are handled better with full transparency of what's happening, i.e. classic implementation, and some of them are way better off with Terrajet because we'd repeat what Terraform has done for them anyway. In addition, the conversion cost is non-zero; we have the conversion webhooks for API changes but each conversion needs to be tested manually to see if the behavior matches or TF is doing unexpected things.

In summary, I see that these providers will go closer to 95% TJ over a long period of time but meanwhile it will be a mix and I don't think we should rush to get to that level because of the costs associated are not worth it for every resource. For example, why would you need to convert VPC in AWS to TJ? All features are there, it's beta and there is almost no issue and it's faster to provision than TJ.

In the light of all these, I proposed the following guidelines (added to the proposal) for resource iterations:
* If you encounter a bug that cannot be solved with Terrajet in a meaningful
  timeframe, consider changing its controller to classic hand-written controller
  and bump the API version.
* If an existing classic CRD (hand-written or ACK-generated) has a bug that
  cannot be fixed easily or has a blocker or we continuously get bugs which
  makes it hard to maintain, consider replacing it with Terrajet and if you do
  bump the API version.

@Feggah Does that clear the confusion for you? Would you look for more concrete conditions?

Reply:
> Does that clear the confusion for you? Would you look for more concrete conditions?

Makes sense, thank you. I thought TJ wasn't the perfect implementation for MRs, but actually it is. No more doubts 👍🏼

Reply:
> In summary, I see that these providers will go closer to 95% TJ over a long period of time but meanwhile it will be a mix and I don't think we should rush to get to that level because of the costs associated are not worth it for every resource. For example, why would you need to convert VPC in AWS to TJ? All features are there, it's beta and there is almost no issue and it's faster to provision than TJ.

This doesn't sound aggressive enough in my opinion. The "if it's not broke don't fix it" approach you mention with VPCs don't resonate with me because it leaves us in a dual implementation where maintainers and contributors need to understand how to work on a native implementation (with all its complexities) for potentially months just because we don't want to commit the time to port it to the more common (TJ) approach and validate that that approach works well.

I'm okay being in that dual implementation state for a significant number of managed resources (~120 MRs in AWS) for a limited time (say 6 months) as long as there's an explicit goal to align as many managed resources as possible on one implementation. I do not feel good about 20+ managed resources using a 'deprecated' implementation indefinitely.

Reply:
I am fine with being as aggressive as possible but there is a limit that's set by our capacity and the nature of contributions. For example, I don't think it's practical to set a timeline like 6 months without having people dedicated to this work, which is quite hard since it's less of a concern for users and contributors but more for maintainers which is a very small set.

I changed the guideline to say `consider converting to Terrajet before fixing the bug` and added a sentence to say that we actively try to converge on Terrajet but still dismissing a planned push that will cost us engineer cycles today.

> We will also update the guidelines about how to add a resource roughly with the
following:
> * If you are adding a new resource, use Terrajet pipeline unless there is a
  blocker.
> * If you encounter a bug that cannot be solved with Terrajet in a meaningful
  timeframe, consider converting its controller to classic hand-written
  controller and bump the API version.
> * If an existing classic CRD (hand-written or ACK-generated) has a bug, you
  should consider converting its controller to Terrajet first.

> The end goal for the conversions is to have more homogeneity in the codebase so
we should convert as many managed resources as possible to Terrajet. Today,
Terrajet is the technology that these three Crossplane providers decide as the
best available for most cases, hence we'll converge on it. Tomorrow, it might be
AWS Cloud Control API for provider-aws, ARM for provider-azure etc. so we'll
have a goal of converging to those technologies. But we won't have a big push to
convert everything to Terrajet immediately because depending on the complexity
of the resource, the conversions require extensive testing to make sure the
behavior matches and it carries risk for existing users.

Reply:
@negz I added the following to the guideline as we discussed offline to make it more aggressive.

> Existing classic CRDs that are on alpha version will be converted to Terrajet before bumping to beta unless there is a blocker.

Also updated the rest of the sections to be inline with converging on Terrajet instead of staying mixed.

Reply:
I appreciate the huge win that `jet` offers in getting resources to be managed by Crossplane. I want to call out some of the non-technical consequences of moving to a Jet as the default technology underlying crossplane providers. Namely, some of the gaps we have already encountered using Jet providers. As adoption increases, crossplane should expect to invest more time and effort in documentation, resource behavior incompatibility, resource interoperability issues and error-handling. 

1. Documentation: Not all Terraform behavior is intuitive. A good example of this can be found on the [google_project_iam](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/google_project_iam) resources. The implications of choosing, for example, `google_project_iam_policy` over `google_project_iam_member` need to be surfaced to users of provider-jet-gcp in the provider-jet-gcp (and not Terraform) docs. [Current Provider-Jet-GCP docs](https://doc.crds.dev/github.com/crossplane-contrib/provider-jet-gcp/cloudplatform.gcp.jet.crossplane.io/ProjectIAMPolicy) do not represent these notes. 
1. Resource Behavior: Some Terraform behavior does not fit nicely into Crossplane's Resource Model. See [this issue](https://github.com/crossplane-contrib/provider-jet-gcp/issues/46) where secret data that _can_ be updated by terraform requires a currently missing event from Kubernetes/Crossplane to get the Update pipeline to run. Jet-providers which need to detect secret updates will require machinery to include this in the "observe" step. Successful migration to a fully jet-based system will require building support around issues like this.
1. Resource Interoperability: Terraform supports several [built-in functions](https://www.terraform.io/language/functions) that ease the friction of chaining provider resources. For providers which require data in a [specific format](https://github.com/crossplane-contrib/provider-jet-vault/issues/1), crossplane providers will need to support this as well. This could be mean extending the transformation capabilities of composition or extending the ability to control a resource's inputs in [terrajet](https://github.com/crossplane/terrajet/issues/260).
1. Error-Handling: Errors output from terraform will need to be complete and easily readable. A recent error encountered by one user in my org looked like the following:
```
accessNotConfigured. If you received a 403 error, make sure you have the `roles/resourcemanager.projectCreator` permission: : File name: main.tf.json
```
This error was incomplete. Rendering the resource in terraform and manually running it revealed the following, which included more debugging information:
```
│ Error: error creating project platform-test-project (Platform Test Project): googleapi: Error 403: Cloud Resource Manager API has not been used in project 295269043414 before or it is disabled. Enable it by visiting https://console.developers.google.com/apis/api/cloudresourcemanager.googleapis.com/overview?project=295269043414 then retry. If you enabled this API recently, wait a few minutes for the action to propagate to our systems and retry.
│ Details:
│ [
│   {
│     "@type": "type.googleapis.com/google.rpc.Help",
│     "links": [
│       {
│         "description": "Google developers console API activation",
│         "url": "https://console.developers.google.com/apis/api/cloudresourcemanager.googleapis.com/overview?project=295269043414"
│       }
│     ]
│   },
│   {
│     "@type": "type.googleapis.com/google.rpc.ErrorInfo",
│     "domain": "googleapis.com",
│     "metadata": {
│       "consumer": "projects/295269043414",
│       "service": "cloudresourcemanager.googleapis.com"
│     },
│     "reason": "SERVICE_DISABLED"
│   }
│ ]
│ , accessNotConfigured. If you received a 403 error, make sure you have the `roles/resourcemanager.projectCreator` permission
│
│   with google_project.platform_test,
│   on main.tf line 7, in resource "google_project" "platform_test":
│    7: resource "google_project" "platform_test" {
```

Another example recently encountered: providing expired GCP credentials to a provider-jet-gcp can result in this error:
```
{"error":"invalid_grant","error_description":"Invalid JWT Signature."}
```

Currently, the only way to debug an error like that is to search the web for "'Invalid JWT Signature' terraform provider gcp"

This is not meant to be a comprehensive list of the issues that terrajet adoption implies. It is simply offered to provide context on how far-reaching this decision will become. Adopting terrajet as the default implementation offers rapid coverage and alleviates the technical burden of building up the library of crossplane-managed resources. At the same time, it creates pressure on downstream support efforts. As the Crossplane Team builds tooling and community around this approach, decisions forced by terrajet adoption will become long-term support requirements for the project.